File: algorithm/sort/CountingSort.java
package com.example.sort;

import java.util.Arrays;
import java.util.List;

public class CountingSort {

    public List<Integer> countingSort(List<Integer> list) {
        int maxNum = getMaxNum(list);
        int[] count = new int[maxNum + 1];
        for (Integer i : list) count[i]++;
        for (int i = 0; i < count.length - 1; i++) count[i + 1] += count[i];
        int[] answer = new int[list.size()];
        for (Integer i : list) {
            count[i] -= 1;
            answer[count[i]] = i;
        }
        return Arrays.stream(answer).boxed().toList();
    }

    private int getMaxNum(List<Integer> list) {
        if (list.isEmpty()) throw new IllegalArgumentException();
        int maxNum = list.get(0);
        for (int i = 1; i < list.size(); i++) {
            if (maxNum < list.get(i)) maxNum = list.get(i);
        }
        return maxNum;
    }

    public int[] countingSort(int[] arr) {
        int[] count = new int[arr.length + 1];
        for (int i : arr) count[i]++;
        for (int i = 0; i < count.length - 1; i++) count[i + 1] += count[i];
        int[] answer = new int[arr.length];
        for (int i : arr) {
            count[i] -= 1;
            answer[count[i]] = i;
        }
        return answer;
    }
}

File: algorithm/sort/CountingSort.kt
package com.example.sort_kotlin

class CountingSort {
    fun sort(arr: IntArray): IntArray {
        if (arr.isEmpty()) {
            return IntArray(0)
        }
        val maxNum = arr.max()
        val newArr = IntArray(maxNum + 1)
        for (num in arr) {
            newArr[num] += 1
        }
        for (i in 1.. newArr.size - 1) {
            newArr[i] += newArr[i - 1]
        }
        val sortedArr = IntArray(arr.size)
        for (i in arr.indices) {
            newArr[arr[i]] -= 1
            sortedArr[newArr[arr[i]]] = arr[i]
        }
        return sortedArr

        // [3 3 5 5 2]
        // [0 0 0 0 0 0]
        // [0 0 1 2 0 2]
        // [0 0 1 3 3 5]
        // [0 0 0 0 0]
        // [0 0 1 2 3 5] -> [0 0 3 0 0]
        // [0 0 1 1 3 5] -> [0 3 3 0 0]
        // [0 0 1 1 3 4] -> [0 3 3 0 5]
        // [0 0 1 1 3 3] -> [0 3 3 5 5]
        // [0 0 0 1 3 3] -> [2 3 3 5 5]
    }
}
